src/Endpoint.ts

```diff
@@ -0,0 +1,186 @@
+import Discord = require("discord.js");
+import fetch from "node-fetch";
+import { fileBackedObject } from "./FileBackedObject";
+import { PersonalSettings, SharedSettings } from "./SharedSettings";
+
+type EndpointName = string;
+type Endpoints = EndpointName[];
+
+interface EndpointsState {
+    endpoints: Endpoints;
+    lastUpdate: Date;
+}
+
+/**
+ *  Posts links to api endpoints
+ *
+ */
+export default class Endpoint {
+    private endpoints: EndpointsState;
+    private baseUrl: string;
+    private maxDistance: number;
+    private aliases: { [key: string]: string[] };
+
+    public constructor(sharedSettings: SharedSettings, endpointFile: string) {
+        console.log("Requested Endpoint extension.");
+        this.baseUrl = sharedSettings.endpoint.baseUrl;
+        this.maxDistance = sharedSettings.endpoint.maxDistance;
+        this.aliases = sharedSettings.endpoint.aliases || {};
+        this.endpoints = fileBackedObject(endpointFile);
+        this.updateEndpoints();
+    }
+
+    public onEndpoint(message: Discord.Message, isAdmin: boolean, command: string, args: string[]) {
+        if (this.endpoints.endpoints === undefined) {
+            message.reply("Sorry, endpoints are not yet initialized!");
+            return;
+        }
+        const argsString = args.reduce((prev, current) => prev + current, "");
+        let minDist = Infinity;
+        let minEndpoint = null;
+
+        for (const endpoint of this.endpoints.endpoints) {
+            const dist = this.calculateEndpointAndSynonymsDist(endpoint, argsString);
+            if (dist < minDist) {
+                minDist = dist;
+                minEndpoint = endpoint;
+            }
+        }
+
+        if (minEndpoint != null && minDist < this.maxDistance) {
+            message.reply(this.baseUrl + minEndpoint);
+        } else {
+            message.reply("Could not find requested endpoint!");
+        }
+    }
+
+    /**
+     * Levenshtein Distance with some simple modifications for endpoint & input
+     * @param endpoint
+     * @param input
+     */
+    private calculateDistance(endpoint: string, input: string): number {
+        endpoint = endpoint.replace(new RegExp("-", "g"), " ");
+        input = input.replace(new RegExp("-", "g"), " ");
+
+        const d = new Array(endpoint.length + 1);
```

Comment:
Is this an array with an expected length?

Reply:
It is because levenshtein distance needs an array of `s1.length` + 1 times `s2.length + 1` for the input strings `s1` and `s2`